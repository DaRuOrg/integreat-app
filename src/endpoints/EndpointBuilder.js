import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'
import StateMapperBuilder from './StateMapperBuilder'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _url
  _mapper
  _stateMapperBuilder
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name
    this._stateMapperBuilder = new StateMapperBuilder(this)

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a url to the builder
   * @param url The url
   * @return {EndpointBuilder} The builder itself
   */
  withUrl (url) {
    this._url = url
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds a state mapper builder to this builder
   * @return {StateMapperBuilder} The state mapper of this builder
   */
  withStateMapper () {
    return this._stateMapperBuilder
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder}  The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._stateMapperBuilder) {
      throw Error('You have to set a state mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._url, this._mapper, this._stateMapperBuilder.build(), this._refetchLogic)
  }
}

export default EndpointBuilder
