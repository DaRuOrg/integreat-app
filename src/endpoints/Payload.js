class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null) {
    this._isFetching = isFetching
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  get isFetching () {
    return this._isFetching
  }

  get data () {
    return this._data
  }

  ready () {
    return !!this._data
  }

  get error () {
    return this._error
  }

  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
