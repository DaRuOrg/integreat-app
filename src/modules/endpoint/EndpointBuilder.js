import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _routerToUrlMapper
  _mapper
  _responseOverride
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a router to url mapper to the builder
   * @param routerToUrlMapper The url
   * @return {EndpointBuilder} The builder itself
   */
  withRouterToUrlMapper (routerToUrlMapper) {
    this._routerToUrlMapper = routerToUrlMapper
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder} The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride) {
    this._responseOverride = responseOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._routerToUrlMapper) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._routerToUrlMapper, this._mapper, this._refetchLogic, this._responseOverride)
  }
}

export default EndpointBuilder
