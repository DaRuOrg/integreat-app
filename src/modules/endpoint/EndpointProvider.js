import React from 'react'
import PropTypes from 'prop-types'
import Endpoint from './Endpoint'
import Error from '../common/containers/Error'

class EndpointProvider extends React.Component {
  static propTypes = {
    endpoints: PropTypes.arrayOf(PropTypes.instanceOf(Endpoint)).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    getEndpoint: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      getEndpoint: (endpointName) => {
        const endpoint = this.endpoints[endpointName]
        if (!endpoint) {
          throw new Error('The endpoint ' + endpointName + ' was not found. Maybe you didn\'t wrap the ' +
            'withFetcher(...) call in a <EndpointProvider /> or the endpoint is not registered')
        }
        return endpoint
      }
    }
  }

  constructor (props, context) {
    super(props, context)
    this.endpoints = props.endpoints.reduce((accumulator, endpoint) => {
      accumulator[endpoint.stateName] = endpoint
      return accumulator
    }, {})
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
