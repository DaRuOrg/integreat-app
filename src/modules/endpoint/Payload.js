/**
 * The payload gets stored in the redux store and holds the information about a etch
 */
class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null, fetchDate = null) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  /**
   * @return {number} The date the fetch was initiated as serializable number
   */
  get fetchDate () {
    return this._fetchDate
  }

  /**
   * @return {boolean} If a fetch is going on
   */
  get isFetching () {
    return this._isFetching
  }

  /**
   * @return {*} The data which has been fetched or null
   */
  get data () {
    return this._data
  }

  /**
   * @return {boolean} If the {@link data} is ready to be used
   */
  ready () {
    return !!this._data
  }

  /**
   * @return {string} The error message if the fetch failed or null
   */
  get error () {
    return this._error
  }

  /**
   * @return {string} The url which was used to initiate the fetch
   */
  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
