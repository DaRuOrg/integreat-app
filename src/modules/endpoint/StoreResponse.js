class StoreResponse {
  /**
   * @return {Promise} The promise which will be resolved in the by the store (more specifically by redux-thunk)
   */
  _promise,
  /**
   * @type {boolean} Whether the correct data is available and ready for the fetcher to be displayed.
   */
  _data_available

  constructor (data_available, promise = Promise.resolve()) {
    this._promise = promise
    this._data_available = data_available
  }
}

export default StoreResponse
