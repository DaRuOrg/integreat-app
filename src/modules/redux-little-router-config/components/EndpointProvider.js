import React from 'react'
import PropTypes from 'prop-types'
import RouteConfig from '../RouteConfig'

class EndpointProvider extends React.Component {
  static propTypes = {
    routeConfig: PropTypes.instanceOf(RouteConfig).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    matchRoute: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      matchRoute: (id) => this.props.routeConfig.matchRoute(id)
    }
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
