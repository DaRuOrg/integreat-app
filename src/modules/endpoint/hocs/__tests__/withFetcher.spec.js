import React from 'react'

import { mount, shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

import EndpointBuilder from '../../EndpointBuilder'
import StoreResponse from '../../StoreResponse'
import Payload from '../../Payload'
import connectedWithFetcher, { withFetcher } from '../withFetcher'
import EndpointProvider from '../../EndpointProvider'

describe('withFetcher', () => {
  const urlParams = {var1: 'a', var2: 'b'}
  const responseOverride = {data: 'random'}
  const endpoint = new EndpointBuilder('endpoint')
    .withUrl('https://someendpoint/{var1}/{var2}/api.json')
    .withMapper((json) => json)
    .withStateMapper().fromFunction(() => (urlParams))
    .withResponseOverride(responseOverride)
    .build()

  // eslint-disable-next-line react/prop-types
  const createComponent = ({endpoint, hideError = false, hideSpinner = false, urlParams = {}, requestAction, classname, otherProps = {[endpoint.payloadName]: new Payload(false)}}) => {
    const HOC = withFetcher(endpoint.stateName, hideError, hideSpinner)

    class WrappedComponent extends React.Component {
      static displayName = 'WrappedComponent'

      render () {
        return <span>WrappedComponent</span>
      }
    }

    const Hoced = HOC(WrappedComponent)

    return <Hoced getEndpoint={() => endpoint}
                  urlParams={urlParams}
                  requestAction={requestAction}
                  classname={classname}
                  {...otherProps} />
  }

  test('should should show error if there is one and it\'s not hidden', () => {
    const hoc = createComponent({
      endpoint,
      hideError: false,
      requestAction: () => new StoreResponse(true),
      otherProps: {[endpoint.payloadName]: new Payload(false, null, 'Yepp... Error time! Wushhh!')}
    })

    expect(shallow(hoc)).toMatchSnapshot()
  })

  test('should should show spinner if there is no data yet and it\'s not hidden', () => {
    const hoc = createComponent({
      endpoint,
      hideSpinner: false,
      requestAction: () => new StoreResponse(false),
      otherProps: {[endpoint.payloadName]: new Payload(true)}
    })

    expect(shallow(hoc)).toMatchSnapshot()
  })

  test('should should show nothing if there is no data yet and spinner is hidden', () => {
    const hoc = createComponent({
      endpoint,
      hideSpinner: true,
      requestAction: () => new StoreResponse(false),
      otherProps: {[endpoint.payloadName]: new Payload(true)}
    })

    expect(shallow(hoc)).toMatchSnapshot()
  })

  test('should should show wrapped component if there is data', () => {
    const hoc = createComponent({
      endpoint,
      requestAction: () => new StoreResponse(true),
      otherProps: {[endpoint.payloadName]: new Payload(false, {})}
    })

    expect(shallow(hoc)).toMatchSnapshot()
  })

  test('should fetch when endpoint tells us', () => {
    const endpoint = new EndpointBuilder('endpoint')
      .withUrl('https://someendpoint/{var1}/{var2}/api.json')
      .withMapper((json) => json)
      .withResponseOverride({})
      .withRefetchLogic(() => true) // Refetch always
      .build()

    const mockRequestAction = jest.fn().mockReturnValue(new StoreResponse(false))
    const hoc = shallow(createComponent({
      endpoint,
      requestAction: mockRequestAction
    }))

    hoc.setProps({...hoc.props()}) // Just call componentWillReceiveProps

    expect(mockRequestAction.mock.calls).toHaveLength(2)
  })

  test('should fetch when props change', () => {
    const mockRequestAction = jest.fn().mockReturnValue(new StoreResponse(false))
    const otherProps = {[endpoint.payloadName]: new Payload(false)}
    const hoc = shallow(createComponent({
      endpoint,
      requestAction: mockRequestAction,
      otherProps
    }))

    hoc.setProps({...hoc.props(), ...otherProps}) // No change in props

    // mockRequestAction has been called once in componentWillMount but it shouldn't have been a second time
    expect(mockRequestAction.mock.calls).not.toHaveLength(2)

    const newProps = {[endpoint.payloadName]: new Payload(false)} // newProps !== otherProps (identity)
    hoc.setProps({...hoc.props(), ...newProps})

    expect(mockRequestAction.mock.calls).toHaveLength(2)
  })

  test('should dispatch the correct actions whens fetch occurs', () => {
    const urlParams = {param1: 'a'}
    const otherUrlParams = {param2: 'b'}
    const mockRequestAction = jest.fn().mockReturnValue(new StoreResponse(true))

    const hoc = shallow(createComponent({endpoint, urlParams, requestAction: mockRequestAction}))
    const instance = hoc.instance()

    expect(hoc.state()).toEqual({isDataAvailable: true})

    expect(mockRequestAction).toBeCalledWith(urlParams)

    instance.fetch(otherUrlParams)

    expect(mockRequestAction).toBeCalledWith(otherUrlParams)

    expect(() => instance.fetch(undefined)).toThrow()
  })

  const mockStore = configureMockStore([thunk])

  describe('connect()', () => {
    test('should map dispatch to props', () => {
      const payload = new Payload(false)
      const store = mockStore({[endpoint.stateName]: payload})
      const HOC = connectedWithFetcher(endpoint.stateName)
      const WrappedComponent = () => <span>WrappedComponent</span>
      const Hoced = HOC(WrappedComponent)

      const tree = mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[endpoint]}>
            <Hoced />
          </EndpointProvider>
        </Provider>
      )

      const wrappedHOCProps = tree.find('endpointFetcher').props()

      expect(store.getActions()).toHaveLength(2) // componentDidMount calls fetch
      expect(store.getActions()).toContainEqual({
        payload: new Payload(false, responseOverride, null, 'https://someendpoint/a/b/api.json', expect.any(Number)),
        type: 'FINISH_FETCH_DATA_ENDPOINT'
      })
      wrappedHOCProps.requestAction({var1: 'c', var2: 'd'})
      expect(store.getActions()).not.toHaveLength(2) // should change after dispatch -> not 2 anymore
      expect(store.getActions()).toContainEqual({
        payload: new Payload(false, responseOverride, null, 'https://someendpoint/c/d/api.json', expect.any(Number)),
        type: 'FINISH_FETCH_DATA_ENDPOINT'
      })
    })

    test('should map state to props', () => {
      /*  The url from the last request. This tells the endpoint that we already have data and displays the wrapped
      component instead of a spinner */
      const requestUrl = 'https://someendpoint/a/b/api.json'
      const payload = new Payload(false, {}, null, requestUrl)
      const store = mockStore({[endpoint.stateName]: payload})
      const HOC = connectedWithFetcher(endpoint.stateName)
      const WrappedComponent = () => <span>WrappedComponent</span>
      const Hoced = HOC(WrappedComponent)

      const tree = mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[endpoint]}>
            <Hoced />
          </EndpointProvider>
        </Provider>
      )

      const wrappedHOCProps = tree.find(WrappedComponent).props()

      expect(wrappedHOCProps).toEqual({
        endpoint: payload.data
      })
    })
  })
})
