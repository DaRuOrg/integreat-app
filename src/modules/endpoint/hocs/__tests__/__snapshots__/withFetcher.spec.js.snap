// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFetcher should should show error if there is one and it's not hidded 1`] = `
<Component
  className="loading"
  error="Yepp... Error time! Wuschhh!"
/>
`;

exports[`withFetcher should should show nothing if there is no data yet and spinner is hidden 1`] = `<div />`;

exports[`withFetcher should should show spinner if there is no data yet and it's not hidden 1`] = `
<Spinner
  className="loading"
  fadeIn="full"
  name="line-scale-party"
  noFadeIn={false}
  overrideSpinnerClassName=""
/>
`;

exports[`withFetcher should should show wrapped component if there is data 1`] = `
<WrappedComponent
  endpoint={Object {}}
  endpointPayload={
    Payload {
      "_data": Object {},
      "_error": null,
      "_fetchDate": null,
      "_isFetching": false,
      "_requestUrl": null,
    }
  }
  requestAction={[Function]}
  urlParams={Object {}}
/>
`;
